### ISSUE TITLE
Enhance matchmaker sub-system

### ISSUE THREAD

**Issue by novabyte** [heroiclabs/nakama#11]
Support more complex fuzzy matches and multi-index search queries on user accounts and storage objects to generate opponent lists for multiplayer. This system will combine the in-memory presence system more directly with the persisted state to create more powerful query/filter composition. This depends on GIN indexes and full-text support on cockroachdb.

**Comment by prodigga** [comment]
Maybe the API could be something like this:

```
var matchFindRequest = new NMatchFindMessage.Builder().
SearchRooms(string[]{"1v1","CaptureTheFlag","SearchAndDestroy"}).
AddSearchParam("MMR",50, 25).
SetTimeout(20).
Build();

client.Send(matchFindRequest , (INMatchMakingResult result) =>
{
  //found match
}, (INError error) => {
  //error
});
```

**_SearchRooms_** - Only match make with people in the same 'room'. This is just a way for us developers to filter out which kind of games the player is going to play in. Users do not 'create' rooms. Maybe developers create 'rooms' via command line on the server, similar to leaderborads? The developer can specify how many users per match in the room, etc. 

In the game, player presses: _Start -> Multiplayer -> Free For All -> Find Match_ so we send an _NMatchFindMessage_ with `SearchRooms(string[]{"FFA"})`. If user selects _Start -> Multiplayer -> All-> Find Match_, we send an _NMatchFindMessage_ with `SearchRooms(string[]{"FFA","DM","CTF"})`. 

_**AddSearchParam**_ - 'fuzzy' match making values. IE Pass in an "MMR" of "50"and a max search range of "25" (ie match against others with an MMR of 25-75. I am not sure if this is the 'best' way to do it, but just an idea.

_**SetTimeout**_ Time out for the search operation. Return an error if match not found after this time?

_**INMatchMakingResult **_ Contains a reference to the INMatch that was found, and also 

```
var cancelMatchFind = new NCancelMatchFindMessage.Default();
client.Send(cancelMatchFind, //etc....
```

To cancel current match find operation?

Just suggestions

**Comment by prodigga** [comment]
Matchmaking should support group play. 

IE I want to play a match together with my friend. I ask nakama to match make for me and my friend (maybe I provide both our UserIDs as well as our "MMR"). The server puts us together in to a match.

I guess it would also be important that the players be out into the same team, so the server not only has to find a game with 2 free spots, but one with 2 free spots on the same team.

**Comment by zyro** [comment]
The initial framework for this feature is available on master since 46e50e1052123dcef13e8dcc3f657d8f99800e91. Matchmaking criteria improvements remain scheduled for 0.14.0.

**Comment by risheelk** [comment]
i like to add filters like range like rank 2- 4 and score 2000 -2800 etc
player will send this to server and server should match other who is also having rank and score within the range

**Comment by mofirouz** [comment]
This is now available on commit https://github.com/heroiclabs/nakama/commit/dc0a5964d8a1fe3dde79b1c6d1d8d0a59b71e216 and will be available in Nakama 1.1.0+
